core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/streams/streams.go

```diff
@@ -178,65 +176,94 @@ func (s *streams) buildResult(ctx context.Context, i int, checkResult ocr2keeper
-		checkResults[i].IneligibilityReason = uint8(reason)
-		return
+		checkResults[i].Retryable = false
+		checkResults[i].PipelineExecutionState = uint8(mercury.PackUnpackDecodeFailed)
+		return err
 	}
 
-	for j, v := range values {
-		s.lggr.Infof("at block %d upkeep %s requested time %s doMercuryRequest values[%d]: %s", lookup.Block, lookup.UpkeepId, lookup.Time, j, hexutil.Encode(v))
+	var mercuryBytes hexutil.Bytes
+	args := map[string]interface{}{
+		"to":   s.registry.Address().Hex(),
+		"data": hexutil.Bytes(payload),
 	}
 
-	state, retryable, mercuryBytes, err := s.checkCallback(ctx, values, lookup)
-	if err != nil {
-		s.lggr.Errorf("at block %d upkeep %s checkCallback err: %s", lookup.Block, lookup.UpkeepId, err.Error())
-		checkResults[i].Retryable = retryable
-		checkResults[i].PipelineExecutionState = uint8(state)
-		return
+	// call checkCallback function at the block which OCR3 has agreed upon
+	if err = s.client.CallContext(ctx, &mercuryBytes, "eth_call", args, hexutil.EncodeUint64(lookup.Block)); err != nil {
+		checkResults[i].Retryable = true
+		checkResults[i].PipelineExecutionState = uint8(mercury.RpcFlakyFailure)
+		return err
 	}
+
 	s.lggr.Infof("at block %d upkeep %s requested time %s checkCallback mercuryBytes: %s", lookup.Block, lookup.UpkeepId, lookup.Time, hexutil.Encode(mercuryBytes))
 
 	unpackCallBackState, needed, performData, failureReason, _, err := s.packer.UnpackCheckCallbackResult(mercuryBytes)
 	if err != nil {
-		s.lggr.Errorf("at block %d upkeep %s requested time %s UnpackCheckCallbackResult err: %s", lookup.Block, lookup.UpkeepId, lookup.Time, err.Error())
 		checkResults[i].PipelineExecutionState = unpackCallBackState
-		return
+		return err
 	}
 
 	if failureReason == uint8(mercury.MercuryUpkeepFailureReasonMercuryCallbackReverted) {
 		checkResults[i].IneligibilityReason = uint8(mercury.MercuryUpkeepFailureReasonMercuryCallbackReverted)
 		s.lggr.Debugf("at block %d upkeep %s requested time %s mercury callback reverts", lookup.Block, lookup.UpkeepId, lookup.Time)
-		return
+		return nil
+
 	}
 
 	if !needed {
 		checkResults[i].IneligibilityReason = uint8(mercury.MercuryUpkeepFailureReasonUpkeepNotNeeded)
 		s.lggr.Debugf("at block %d upkeep %s requested time %s callback reports upkeep not needed", lookup.Block, lookup.UpkeepId, lookup.Time)
-		return
+		return nil
 	}
 
 	checkResults[i].IneligibilityReason = uint8(mercury.MercuryUpkeepFailureReasonNone)
 	checkResults[i].Eligible = true
 	checkResults[i].PerformData = performData
 	s.lggr.Infof("at block %d upkeep %s requested time %s successful with perform data: %s", lookup.Block, lookup.UpkeepId, lookup.Time, hexutil.Encode(performData))
+
+	return nil
 }
 
-// allowedToUseMercury retrieves upkeep's administrative offchain config and decode a mercuryEnabled bool to indicate if
+func (s *streams) DoMercuryRequest(ctx context.Context, lookup *mercury.StreamsLookup, checkResults []ocr2keepers.CheckResult, i int) ([][]byte, error) {
```

Comment:
For the scope of this PR, I kept the "index and slice" change out of the scope of this PR. It is just modularize the functions, and update the debugging script using modular functions. I hope this can make the code safer and easier to review.

We can revisit the index/slice changes in following, smaller PRs. Hope that is okay.

Reply:
of course :) thanks for the explanation